library/nist_kw.c

```diff
@@ -0,0 +1,756 @@
+        ret = mbedtls_cipher_update( &ctx->cipher_ctx,
+                                     inbuff, 16, output, &olen );
+        if( ret != 0 )
+            goto cleanup;
+    }
+    else
+    {
+        /*
+         * Do the wrapping function W, as defined in RFC 3394 section 2.2.1
+         */
+        if( semiblocks < MIN_SEMIBLOCKS_COUNT )
+        {
+            ret = MBEDTLS_ERR_CIPHER_BAD_INPUT_DATA;
+            goto cleanup;
+        }
+
+        /* Calculate intermediate values */
+        for( t = 1; t <= s; t++ )
+        {
+            memcpy( inbuff, A, KW_SEMIBLOCK_LENGTH );
+            memcpy( inbuff + KW_SEMIBLOCK_LENGTH, R2, KW_SEMIBLOCK_LENGTH );
+
+            ret = mbedtls_cipher_update( &ctx->cipher_ctx,
+                                         inbuff, 16, outbuff, &olen );
+            if( ret != 0 )
+                goto cleanup;
+
+            memcpy( A, outbuff, KW_SEMIBLOCK_LENGTH );
+            calc_a_xor_t( A, t );
+
+            memcpy( R2, outbuff + KW_SEMIBLOCK_LENGTH, KW_SEMIBLOCK_LENGTH );
+            R2 += KW_SEMIBLOCK_LENGTH;
+            if( R2 >= output + ( semiblocks * KW_SEMIBLOCK_LENGTH ) )
+                R2 = output + KW_SEMIBLOCK_LENGTH;
+        }
+    }
+    /*
+     * Set the output length as the number.
+     * If mode is MBEDTLS_KW_MODE_KWP and in_len < 8,
+     * Then after padding, the number of semiblocks in the output would be 2,
+     * Which means the output is 16 bytes, as expected.
+     */
+    *out_len = semiblocks * KW_SEMIBLOCK_LENGTH;
+
+cleanup:
+
+    if( ret != 0)
+    {
+        memset( output, 0, semiblocks * KW_SEMIBLOCK_LENGTH );
+    }
+    mbedtls_platform_zeroize( inbuff, KW_SEMIBLOCK_LENGTH * 2 );
+    mbedtls_platform_zeroize( outbuff, KW_SEMIBLOCK_LENGTH * 2 );
+    mbedtls_cipher_finish( &ctx->cipher_ctx, NULL, &olen );
+    return( ret );
+}
+
+/*
+ * W-1 function as defined in RFC 3394 section 2.2.2
+ * This function assumes output buffer is at least of size
+ * ( semiblocks - 1 ) * KW_SEMIBLOCK_LENGTH , and the the input buffer
+ * is of size semiblocks * KW_SEMIBLOCK_LENGTH. A is the first semiblock of the
+ * input buffer given by the user.
+ */
+static int unwrap( mbedtls_nist_kw_context *ctx,
+                   const unsigned char *input, size_t semiblocks,
+                   unsigned char A[KW_SEMIBLOCK_LENGTH],
+                   unsigned char *output, size_t* out_len )
+{
+    int ret = 0;
+    const size_t s = 6 * ( semiblocks - 1 );
+    size_t olen;
+    uint64_t t = 0;
+    unsigned char outbuff[KW_SEMIBLOCK_LENGTH * 2];
+    unsigned char inbuff[KW_SEMIBLOCK_LENGTH * 2];
+    unsigned char *R = output + ( semiblocks - 2 ) * KW_SEMIBLOCK_LENGTH;
+    *out_len = 0;
+
+    if( semiblocks < MIN_SEMIBLOCKS_COUNT )
+    {
+        return( MBEDTLS_ERR_CIPHER_BAD_INPUT_DATA );
+    }
+
+    memcpy( A, input, KW_SEMIBLOCK_LENGTH );
+    memmove( output, input + KW_SEMIBLOCK_LENGTH, ( semiblocks - 1 ) * KW_SEMIBLOCK_LENGTH );
```

Comment:
This can overflow if the output buffer capacity is smaller than `( semiblocks - 1 ) * KW_SEMIBLOCK_LENGTH`.

Reply:
But this is in the internal static function `unwrap`, which is called by `mbedtls_nist_kw_unwrap()`. The output buffer size is checked in the public API, so the size of the output buffer in this function is guaranteed to be at least of size ( semiblocks - 1 ) * KW_SEMIBLOCK_LENGTH.

Reply:
`mbedtls_nist_kw_unwrap()` is checking for `*out_len >= in_len - KW_SEMIBLOCK_LENGTH`, but not `*out_len >= KW_SEMIBLOCK_LENGTH`, which might happen if in_len is too small as it is in some of your tests, like the "KW ciphertext too short" ones.

Reply:
Sorry, nevermind. I had a closer look and this is OK.

Reply:
for the record, `in_len` is guranteed to be `in_len >= KW_SEMIBLOCK_LENGTH * 2` for KWP and `in_len` >= 24 for KW.
In addition, if `*out_len >= in_len - KW_SEMIBLOCK_LENGTH` , and `semiblocks` is the size of in_len in semiblocks, then `in_len - KW_SEMIBLOCK_LENGTH == (semiblocks - 1 ) * KW_SEMIBLOCK_LENGTH`